devito/passes/iet/mpi.py

```diff
@@ -59,120 +59,116 @@ def _hoist_halospots(iet):
+        for i, hs0 in enumerate(halo_spots):
 
-                for n, i in enumerate(iters):
-                    if i not in scopes:
-                        continue
+            # Nothing to do if the HaloSpot is void
+            if hs0.halo_scheme.is_void:
+                continue
 
-                    candidates = [i.dim._defines for i in iters[n:]]
+            for hs1 in halo_spots[i+1:]:
+                # If there are Conditionals involved, both `hs0` and `hs1` must be
+                # within the same Conditional, otherwise we would break the control
+                if cond_mapper.get(hs0) != cond_mapper.get(hs1):
+                    continue
 
-                    all_candidates = set().union(*candidates)
-                    reads = scopes[i].getreads(f)
-                    if any(set(a.ispace.dimensions) & all_candidates
-                           for a in reads):
-                        continue
+                # If there are overlapping time accesses, skip
+                if any(i in hs0.halo_scheme.loc_values
+                        for i in hs1.halo_scheme.loc_values):
+                    continue
 
-                    for dep in scopes[i].d_flow.project(f):
-                        if not any(r(dep, candidates, loc_dims) for r in rules):
-                            break
-                    else:
-                        hsmapper[hs] = hsmapper[hs].drop(f)
-                        imapper[i].append(hs.halo_scheme.project(f))
-                        break
+                # Compare hs0 to subsequent halo_spots, looking for optimization
+                # possibilities
+                _process_halo_to_halo(hs0, hs1, iters, scopes, hsmapper, imapper)
 
-    # Post-process analysis
     mapper = {i: HaloSpot(i._rebuild(), HaloScheme.union(hss))
               for i, hss in imapper.items()}
     mapper.update({i: i.body if hs.is_void else i._rebuild(halo_scheme=hs)
                    for i, hs in hsmapper.items()})
 
-    # Transform the IET hoisting/dropping HaloSpots as according to the analysis
-    iet = Transformer(mapper, nested=True).visit(iet)
-
-    # Clean up: de-nest HaloSpots if necessary
-    mapper = {}
-    for hs in FindNodes(HaloSpot).visit(iet):
-        if hs.body.is_HaloSpot:
-            halo_scheme = HaloScheme.union([hs.halo_scheme, hs.body.halo_scheme])
-            mapper[hs] = hs._rebuild(halo_scheme=halo_scheme, body=hs.body.body)
     iet = Transformer(mapper, nested=True).visit(iet)
 
     return iet
 
 
-def _merge_halospots(iet):
-    """
-    Merge HaloSpots on the same Iteration tree level where all data dependencies
-    would be honored.
-    """
+def _process_halo_to_halo(hs0, hs1, iters, scopes, hsmapper, imapper):
 
-    # Merge rules -- if the retval is True, then it means the input `dep` is not
-    # a stopper to halo merging
+    # Loop over the functions in the HaloSpots
+    for f, v in hs1.fmapper.items():
+        # If no time accesses, skip
```

Comment:
"E.g., if no `time` accesses, skip"

you have to say it's an example...

Reply:
wdym by 'say it's an example' ?